nca/CoreDS/ConnectivityProperties.py

```diff
@@ -0,0 +1,587 @@
+        return real_ports
+
+    @staticmethod
+    def make_conn_props(peer_container, src_ports=PortSet(True), dst_ports=PortSet(True),
+                        protocols=ProtocolSet(True), src_peers=None, dst_peers=None,
+                        paths_dfa=None, hosts_dfa=None, methods=MethodSet(True),
+                        icmp_type=DimensionsManager().get_dimension_domain_by_name('icmp_type'),
+                        icmp_code=DimensionsManager().get_dimension_domain_by_name('icmp_code')):
+        """
+        get ConnectivityProperties with allowed connections, corresponding to input properties cube.
+        ConnectivityProperties should not contain named ports: substitute them with corresponding port numbers, per peer
+        :param PeerContainer peer_container: The set of endpoints and their namespaces
+        :param PortSet src_ports: ports set for src_ports dimension (possibly containing named ports)
+        :param PortSet dst_ports: ports set for dst_ports dimension (possibly containing named ports)
+        :param ProtocolSet protocols: CanonicalIntervalSet obj for protocols dimension
+        :param PeerSet src_peers: the set of source peers
+        :param PeerSet dst_peers: the set of target peers
+        :param MinDFA paths_dfa: MinDFA obj for paths dimension
+        :param MinDFA hosts_dfa: MinDFA obj for hosts dimension
+        :param MethodSet methods: CanonicalIntervalSet obj for methods dimension
+        :param CanonicalIntervalSet icmp_type: ICMP-specific parameter (type dimension)
+        :param CanonicalIntervalSet icmp_code: ICMP-specific parameter (code dimension)
+        :return: ConnectivityProperties with allowed connections, corresponding to input properties cube
+        """
+        base_peer_set = peer_container.peer_set.copy()
+        if src_peers:
+            base_peer_set |= src_peers
+            src_peers_interval = base_peer_set.get_peer_interval_of(src_peers)
+        else:
+            src_peers_interval = None
+        if dst_peers:
+            base_peer_set |= dst_peers
+            dst_peers_interval = base_peer_set.get_peer_interval_of(dst_peers)
+        else:
+            dst_peers_interval = None
+        if (not src_ports.named_ports or not src_peers) and (not dst_ports.named_ports or not dst_peers):
+            # Should not resolve named ports
+            return ConnectivityProperties(source_ports=src_ports, dest_ports=dst_ports,
+                                          protocols=protocols, methods=methods, paths=paths_dfa, hosts=hosts_dfa,
+                                          icmp_type=icmp_type, icmp_code=icmp_code,
+                                          src_peers=src_peers_interval, dst_peers=dst_peers_interval,
+                                          base_peer_set=base_peer_set)
+        # Resolving named ports
+        conn_properties = ConnectivityProperties.make_empty_props(peer_container)
+        if src_ports.named_ports and dst_ports.named_ports:
```

Comment:
what about `excluded_named_ports` of `PortSet` objects? they should also be handled and resolved.

Reply:
excluded_named_ports should not appear in ConnectivityProperties together with dst_peers, because:
in the original implementation dst_peers are not a part of ConnectivityProperties and so cannot be resolved in ConnectivityProperties.
In the optimized implementation named ports are resolved as soon as ConnectivityProperties have been constructed, so excluded_named_ports are never created.
Added an assertion to ensure this.

Reply:
I meant that it should be resolved while constructing the `ConnectivityProperties` object.
Adding another comment in `CalicoPolicyYamlParser.py` where this is relevant as an example.

Reply:
Right. Fixed dealing with `excluded_named_ports `- maintaining them in `ConnectivityCube ` and resolving when constructing `ConnectivityProperties `object.